### ISSUE TITLE
False positive for SC2119/SC2120 when using source with arguments in a function

### ISSUE THREAD

**Issue by fvanderveen** [koalaman/shellcheck#869]
- Rule Id (if any, e.g. SC1000): SC2119/SC2120
- My shellcheck version (`shellcheck --version` or "online"): version: 0.4.4 (ubuntu 16.10 apt-get installed)
- [ ] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

Cannot check this online, as shellcheck.net does not have the capability for multiple files + source. However, decided to clone this repo and build locally to verify ("0.4.6"; built from `f0e0d9ff`) and it still happened there.

#### Here's a snippet or screenshot that shows the problem:

bug.sh:
```sh

#!/bin/bash

function bugtrigger() {
    source "bugcheck.sh" "Hello" "World"
    return 0
}

bugtrigger

```

bugcheck.sh:
```sh

#!/bin/bash
echo "$*"

```

#### Here's what shellcheck currently says:

shellcheck bug.sh bugcheck.sh 

In bug.sh line 3:
function bugtrigger() {
^-- SC2120: bugtrigger references arguments, but none are ever passed.

In bug.sh line 8:
bugtrigger
^-- SC2119: Use bugtrigger "$@" if function's $1 should mean script's $1.

#### Here's what I wanted or expected to see:

Neither check should trigger, as the "$*" (or $1, $2, $@, etc) used by the sourced script are already provided by the source call.

Even better would be if the checks would trigger on a source call to such a file _without_ arguments (as if it was a function call), as having the line `source "bugcheck.sh"` in `bug.sh` should - or could - technically trigger the same or similar checks, as the `bugcheck.sh` file uses arguments that would not be passed by the source call.

**Comment by MechaLynx** [comment]
I think I've encountered a variant of this. I was about to open an issue but I'm not sure if it's actually a different bug:

```bash
#! /usr/bin/env bash
# shellcheck disable=2034

O_testvar="foo"
O_testvar2="bar"

function test {
    for v in ${!O_*}
    do
        echo "$v"
    done
}

test
```

I'm using shellcheck 0.4.6 on windows 7, installed through cabal (was using 0.4.5 when I found this, just updated - online version produces the same error).

shellcheck considers this a reference to positional parameters, but it's just an indirection expansion that expands into all variable names prefixed with `O_` and the code behaves as intended. The prefix used in the expansion doesn't seem to make a difference, from the little I've tested.

**Comment by koalaman** [comment]
Both great bugs! They have been fixed in fd79e80.

Like you say, this should probably be a thing for included files as well, but that functionality is not yet there.

**Comment by deomorxsy** [comment]
Awk syntax still raises this with heredocs:

```
core_routine() {  ● core_routine references arguments, but none are ever passed.
(
cat <<EOF
for f in /bin/* do
    case $f in
    */zzzzz) ldd "$(readlink -f "$(which "$f")")" \
        | awk '{print \$3}' >>/foo.txt ;;
    esac
done 
EOF
) | tee /app/test.txt

}
```
A workaround would be to escape the "$3" reference:

```
core_routine() {
(
cat <<EOF
for f in /bin/* do
    case $f in
    */zzzzz) ldd "$(readlink -f "$(which "$f")")" \
        | awk '{print \$3}' >>/foo.txt ;;
    esac
done 
EOF
) | tee /app/test.txt
}
```

The problem is that other programs such as sed recognizes the "\$3" escaped as the error ```sed: unmatched '/'```.
